### PR TITLE
Memory Profiler

### PR DESCRIPTION
**Pull request by tbennun**
This callback provides an itemized profile of where memory is being used throughout an LBANN model.

Other things in this PR:
- [x] Standardize callback behavior w.r.t. checkpoints
- [x] ROCm support for GPU memory usage

At this stage, the callback will *not* gather information across ranks (e.g., for maximal usage). It is also not able to accurately say what is the expected maximal resident memory (because some activations and error signals should be kept), nor anticipate workspace (such as when calling DNN libraries). These will be left for a potential v2.

### CODE REVIEW

File: src/callbacks/memory_profiler.cpp [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+         << "). Size: " << allocated / 1048576.0 << " MiB" << std::endl;
+  return allocated;
+}
+
+memory_profiler::memory_profiler(bool detailed_first_step)
+  : callback_base(), m_detailed_first_step(detailed_first_step)
+{
+#ifndef LBANN_HAS_GPU
+  LBANN_WARNING("Memory profiler callback only provides unaccounted memory "
+                "information with GPU support.");
```
Comment (ndryden) [inline src/callbacks/memory_profiler.cpp:107]: I am having trouble parsing this warning text. Maybe rephrase?

**Review by ndryden — APPROVED** [review]
👍

File: src/callbacks/gpu_memory_usage.cpp [hunk 2]
```diff
@@ -78,10 +78,14 @@ void gpu_memory_usage::write_specific_proto(lbann_data::Callback& proto) const
...
 {
-#ifdef LBANN_HAS_CUDA
+#ifdef LBANN_HAS_GPU
   size_t available;
   size_t total;
+#ifdef LBANN_HAS_CUDA
   FORCE_CHECK_CUDA(cudaMemGetInfo(&available, &total));
+#elif defined(LBANN_HAS_ROCM)
+  FORCE_CHECK_ROCM(hipMemGetInfo(&available, &total));
+#endif
```
Comment (benson31) [inline src/callbacks/gpu_memory_usage.cpp:88]: Could just be
```
auto const [available, total] = h2::gpu::mem_info();
```

File: src/callbacks/memory_profiler.cpp [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+static inline size_t get_used_gpu_memory()
+{
+#ifdef LBANN_HAS_GPU
+  size_t available;
+  size_t total;
+#ifdef LBANN_HAS_CUDA
+  FORCE_CHECK_CUDA(cudaMemGetInfo(&available, &total));
+#elif defined(LBANN_HAS_ROCM)
+  FORCE_CHECK_ROCM(hipMemGetInfo(&available, &total));
+#endif
```
Comment (benson31) [inline src/callbacks/memory_profiler.cpp:58]:
```suggestion
auto const [available, total] = h2::gpu::mem_info();
```
See [H2 GPU memory helpers](https://github.com/LLNL/DiHydrogen/blob/develop/include/h2/gpu/memory_utils.hpp). I expect these to be stable.

File: src/callbacks/memory_profiler.cpp [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+  return total;
```
Comment (benson31) [inline src/callbacks/memory_profiler.cpp:81]:
```suggestion
  return h2::gpu::mem_info().total;
```

**Review by benson31 — APPROVED** [review]
Overall looks good. Thanks for adding this.

File: src/callbacks/memory_profiler.cpp (hunk 1 above)
Comment (tbennun) [reply src/callbacks/memory_profiler.cpp:107]: Fixed, thanks.
